### PR TITLE
Add GSN support for erc20-to-native bridge mode

### PR DESCRIPTION
**Pull request by LyzeOfKiel**
Implementation of token paymaster contract for GSN, added modified version of `executeSignatures` which interacts with paymaster.
**NOTE**: Before starting tests don't forget to start GSN on a local test network via `npx gsn start`.
**ISSUES**: There are two hard-coded constants in the paymaster contract related to its gas consumption. We need to find a way to make them more precise or remove them at all.

### CODE REVIEW

**Review by k1rill-fedoseev — CHANGES_REQUESTED** [review]

File: contracts/upgradeable_contracts/erc20_to_erc20/GSNForeignBridgeErcToErc.sol [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            require(maxTokensFee <= amount);
+            require(contractAddress == address(this));
+            require(!relayedMessages(txHash));
+            setRelayedMessages(txHash, true);
+
+            addTotalExecutedPerDay(getCurrentDay(), amount);
+
+            // Send maxTokensFee to paymaster
+            uint256 unshiftMaxFee = _unshiftValue(maxTokensFee);
+            require(erc20token().transfer(addressStorage[PAYMASTER], unshiftMaxFee), "paymaster tranfer failed");
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/upgradeable_contracts/erc20_to_erc20/GSNForeignBridgeErcToErc.sol:40]: For now I don't see a way to do this. When forwarder calls the target contract it only provides actual function parameters and the address of the message sender (https://github.com/opengsn/gsn/blob/master/contracts/forwarder/Forwarder.sol#L63).

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        return uint256(tmp);
+    }
+
+    function preRelayedCall(
+        GsnTypes.RelayRequest relayRequest,
+        bytes signature,
+        bytes approvalData,
+        uint256 maxPossibleGas
+    ) public returns (bytes memory context, bool revertOnRecipientRevert) {
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol:75]: Fixed

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        (signature, approvalData, maxPossibleGas);
+        // Get 3rd argument of executeSignaturesGSN, i.e. maxTokensFee
+        uint256 maxTokensFee = readUint256(relayRequest.request.data, 4 + 32 + 32);
+        uint256 potentialEthIncome = router.getAmountsOut(maxTokensFee, TokenWethPair)[1];
+        uint256 gasPrice = relayRequest.relayData.gasPrice;
+        uint256 averageGasUsage = 400000;
+        uint256 ethfee = gasPrice * averageGasUsage;
+        require(potentialEthIncome >= ethfee, "tokens fee can't cover expenses");
+        return (abi.encode(relayRequest.request.from), false);
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol]: Fixed

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    function postRelayedCall(bytes context, bool success, uint256 gasUseWithoutPost, GsnTypes.RelayData relayData)
+        public
+    {
+        (success);
+
+        // Calculate commission
+        uint256 chargeWei = relayHub.calculateCharge(gasUseWithoutPost + 194000, relayData);
+
+        // Uniswap
+        uint256 tokenBalanceBefore = erc20().balanceOf(address(this));
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol]: Yes, now I pass `maxTokensFee` to `postRelayedCall` via `context`

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        router.swapTokensForExactETH(chargeWei, tokenBalanceBefore, TokenWethPair, address(this), block.timestamp);
+
+        // Extract destination address from context
+        bytes memory localContext = context;
+        address to;
+        assembly {
+            to := mload(add(localContext, 32))
+        }
+        // Send rest of tokens to user
+        uint256 tokensLeft = erc20().balanceOf(address(this));
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol]: Fixed

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        require(erc20().transfer(to, tokensLeft));
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol]: Agree

File: contracts/mocks/UniswapRouterMock.sol [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    function swapTokensForExactETH(uint256 amountOut, uint256 amountInMax, address[] path, address to, uint256 deadline)
+        external
+        returns (uint256[] memory amounts)
+    {
+        uint256 ethToSend = amountOut;
+        uint256 tokensToTake = ethToSend * 100;
+
+        ERC20(path[0]).transferFrom(msg.sender, address(this), tokensToTake);
+
+        TokenPaymaster(msg.sender).deposit.value(ethToSend)();
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/mocks/UniswapRouterMock.sol]: Done

File: contracts/upgradeable_contracts/erc20_to_erc20/GSNForeignBridgeErcToErc.sol [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+contract GSNForeignBridgeErcToErc is BasicForeignBridgeErcToErc, ERC20Bridge, BaseRelayRecipient {
+    bytes32 internal constant PAYMASTER = 0xfefcc139ed357999ed60c6a013947328d52e7d9751e93fd0274a2bfae5cbcb12; // keccak256(abi.encodePacked("paymaster"))
+
+    function versionRecipient() external view returns (string memory) {
+        return "1.0.1";
+    }
+
+    function setTrustedForwarder(address _trustedForwarder) public onlyOwner {
+        trustedForwarder = _trustedForwarder;
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/upgradeable_contracts/erc20_to_erc20/GSNForeignBridgeErcToErc.sol]: Had to change `BaseRelayRecipient` structure a little for this

File: test/erc_to_erc/gsn/gsn.test.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        [homeDailyLimit, homeMaxPerTx],
+        owner,
+        decimalShiftZero
+      )
+      await foreignBridge.setTrustedForwarder(ForwarderAddress)
+      await foreignBridge.setPayMaster(paymaster.address)
+
+      await token.mint(foreignBridge.address, BRIDGE_TOKENS)
+    })
+    it('should allow to executeSignaturesGSN', async () => {
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply test/erc_to_erc/gsn/gsn.test.js]: Done

File: contracts/gsn/token_paymaster/TokenPaymaster.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol]: Fixed

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+contract TokenPaymaster is BasePaymaster {
+    address private token;
+
+    IUniswapV2Router02 private router;
+    address[] private TokenWethPair = new address[](2);
+
+    constructor(address rh, address fw) public {
+        setRelayHub(IRelayHub(rh));
+        setTrustedForwarder(IForwarder(fw));
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply contracts/gsn/token_paymaster/TokenPaymaster.sol:19]: Done

File: test/erc_to_native/foreign_bridge_gsn.test.js [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+  relayer.addAccount(GSNUser.privateKey)
+  return GSNUser.address
+}
+
+function getEthersGSNContract(truffleContract, address, relayer, from) {
+  const pr = new ethers.providers.Web3Provider(relayer)
+  return new ethers.Contract(address, truffleContract.abi, pr.getSigner(from))
+}
+
+contract('ForeignBridge_ERC20_to_ERC20_GSN', async accounts => {
```
Comment (k1rill-fedoseev) [inline test/erc_to_native/foreign_bridge_gsn.test.js:45]:
```suggestion
contract('ForeignBridge_ERC20_to_Native_GSN', async accounts => {
```

File: contracts/gsn/token_paymaster/TokenPaymaster.sol [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    ) public returns (bytes memory context, bool revertOnRecipientRevert) {
+        (signature, approvalData);
+        _verifyForwarder(relayRequest);
+        bytes memory reqData = relayRequest.request.data;
+        // Verify that `executeSignaturesGSN` is called
+        bytes4 functionSelector;
+        assembly {
+            functionSelector := mload(add(reqData, 32))
+        }
+        require(bridge == relayRequest.request.to, "accepets only bridge calls");
```
Comment (k1rill-fedoseev) [inline contracts/gsn/token_paymaster/TokenPaymaster.sol:91]:
```suggestion
        require(bridge == relayRequest.request.to, "accepts only bridge calls");
```

File: scripts/test.sh [hunk 13]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply scripts/test.sh:1]: Forgot to drop it after rebase

**Review by k1rill-fedoseev — APPROVED** [review]

**Pull request merged**